app/abci/handlers.go

```diff
@@ -0,0 +1,222 @@
+		signature, err := h.signer.Sign(batch.BatchId, utils.Secp256k1)
+		if err != nil {
+			return nil, err
+		}
+
+		h.logger.Debug("submitting batch signature", "signature", signature)
+		return &abci.ResponseExtendVote{VoteExtension: signature}, nil
+	}
+}
+
+// VerifyVoteExtensionHandler handles the VerifyVoteExtension ABCI to
+// verify the batch signature included in the pre-commit vote against
+// the public key registered in the pubkey module.
+func (h *Handlers) VerifyVoteExtensionHandler() sdk.VerifyVoteExtensionHandler {
+	return func(ctx sdk.Context, req *abci.RequestVerifyVoteExtension) (*abci.ResponseVerifyVoteExtension, error) {
+		h.logger.Debug("start verify vote extension handler", "request", req)
+
+		batch, err := h.batchingKeeper.GetBatchForHeight(ctx, ctx.BlockHeight()-1)
+		if err != nil {
+			return nil, err
+		}
+		ok, err := h.verifyBatchSignature(ctx, batch.BatchId, req.VoteExtension[:64], req.ValidatorAddress, utils.Secp256k1)
+		if !ok || err != nil {
+			return nil, err
+		}
+
+		h.logger.Debug("successfully verified signature", "request", req.ValidatorAddress, "height", req.Height)
+		return &abci.ResponseVerifyVoteExtension{Status: abci.ResponseVerifyVoteExtension_ACCEPT}, nil
+	}
+}
+
+// PrepareProposalHandler handles the PrepareProposal ABCI to inject
+// a canonical set of vote extensions in the proposal.
+func (h *Handlers) PrepareProposalHandler() sdk.PrepareProposalHandler {
+	return func(ctx sdk.Context, req *abci.RequestPrepareProposal) (*abci.ResponsePrepareProposal, error) {
+		h.logger.Debug("start prepare proposal handler")
+
+		err := baseapp.ValidateVoteExtensions(ctx, h.stakingKeeper, req.Height, ctx.ChainID(), req.LocalLastCommit)
+		if err != nil {
+			return nil, err
+		}
+
+		proposalTxs := req.Txs
+		if req.Height > ctx.ConsensusParams().Abci.VoteExtensionsEnableHeight {
+			bz, err := json.Marshal(req.LocalLastCommit)
+			if err != nil {
+				h.logger.Error("failed to marshal extended votes", "err", err)
+			}
+			proposalTxs = append([][]byte{bz}, proposalTxs...)
+		}
+
+		return &abci.ResponsePrepareProposal{
+			Txs: proposalTxs,
+		}, nil
+	}
+}
+
+// ProcessProposalHandler handles the ProcessProposal ABCI to validate
+// the canonical set of vote extensions injected by the proposer.
+func (h *Handlers) ProcessProposalHandler() sdk.ProcessProposalHandler {
+	return func(ctx sdk.Context, req *abci.RequestProcessProposal) (*abci.ResponseProcessProposal, error) {
+		if req.Height > ctx.ConsensusParams().Abci.VoteExtensionsEnableHeight {
+			var extendedVotes abci.ExtendedCommitInfo
+			if err := json.Unmarshal(req.Txs[0], &extendedVotes); err != nil {
+				h.logger.Error("failed to decode injected extended votes tx", "err", err)
+				return &abci.ResponseProcessProposal{Status: abci.ResponseProcessProposal_REJECT}, nil
+			}
+
+			// Validate signatures and perform 2/3 voting power check.
+			err := baseapp.ValidateVoteExtensions(ctx, h.stakingKeeper, req.Height, ctx.ChainID(), extendedVotes)
```

Comment:
Where is `ValidateVoteExtensions` and how does it do the 2/3rds check? 😬

Reply:
It's a helper function provided by Cosmos https://github.com/cosmos/cosmos-sdk/blob/8bfcf554275c1efbb42666cc8510d2da139b67fa/baseapp/abci_utils.go#L44

I got a question from re-reading this code. It seems the voting power is not included in the sign message? Not sure how we can trust the power included in the vote if this is the case.

Reply:
Aaah, I thought the comment referred to our 2/3rds check for the batch, but if this code verifies that the received votes all have valid extensions and total more than 2/3rds it should also work for our batches.